src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -138,8 +138,8 @@ public static unsafe void Copy(Array sourceArray, Array destinationArray, int le
             MethodTable* pMT = RuntimeHelpers.GetMethodTable(sourceArray);
             if (pMT == RuntimeHelpers.GetMethodTable(destinationArray) &&
                 !pMT->IsMultiDimensionalArray &&
-                (uint)length <= (nuint)sourceArray.LongLength &&
-                (uint)length <= (nuint)destinationArray.LongLength)
+                (uint)length <= sourceArray.NativeLength &&
```

Comment:
I like the `NativeLength` property to make these casts go away.

Reply:
Part of this was trying to reduce codegen by having a qword comparison directly against the NativeLength field rather than first requiring zero-extension from a dword field into a qword register. Might have to give up that optimization if we can't assume the padding is all-zero, though.

Reply:
I do not think it is worth it to complicate the code to save a few bytes of codegen. It is very x86/64 specific code-size micro-optimization.

Reply:
We could just have `NativeLength` be `=> Unsafe.As<RawArrayData>(this).Length` to force the zero-extension to `nuint`.

That would make it easier to switch to properly be `NativeLength` if we ever decide to make the switch in the future and would avoid the potential problems here while keeping things as is (not requiring more complex changes to actually track `nuint` in the Array metadata today).